File: March-2023/C++/SortAnArray.cxx
// 912. Sort an Array
// Counting sort, get the min and Max range and keep checking if element exists in map
class Solution {
public:
    vector<int> sortArray(vector<int>& nums) {
        unordered_map<int,int> map;
        for(int &n: nums){
            map[n]++;
        }
        // get minimum and maximum 
        int mini = *min_element(nums.begin(),nums.end());
        int maxi = *max_element(nums.begin(),nums.end());
        int j=0;
        for(int i=mini; i<=maxi; i++){
           // To handle duplicates
            while(map[i]>0){
                nums[j]=i;
                j++;
                map[i]--;
                
            }
        }
        return nums;
    }
};

File: March-2023/Java/SortAnArray.java
// 912. Sort an Array
// Counting Sort
class Solution {
    public int[] sortArray(int[] nums) {
      HashMap<Integer,Integer> map = new HashMap<>();
      for(int i=0; i<nums.length; i++){
          map.put(nums[i],map.getOrDefault(nums[i],0)+1);
          
       }
        // get min and Max to use as range in counting sort
        int mini = Arrays.stream(nums).min().getAsInt();
        int maxi = Arrays.stream(nums).max().getAsInt();
        int j=0;
        for(int i=mini; i<=maxi; i++){
            
            while(map.containsKey(i) && map.get(i)>0){
                // Reduce count of duplicates
                map.put(i,map.get(i)-1);
                nums[j]=i;
                j++;
            }
            
        }
        return nums;
    }
}
